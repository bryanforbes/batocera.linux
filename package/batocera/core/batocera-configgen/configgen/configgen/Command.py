from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
