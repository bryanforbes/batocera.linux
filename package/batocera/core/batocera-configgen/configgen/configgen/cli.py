from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
