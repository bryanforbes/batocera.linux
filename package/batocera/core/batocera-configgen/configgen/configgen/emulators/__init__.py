from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
