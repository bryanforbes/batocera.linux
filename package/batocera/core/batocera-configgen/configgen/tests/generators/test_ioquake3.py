from __future__ import annotations

import filecmp
from pathlib import Path
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.config import SystemConfig
from configgen.generators.ioquake3.ioquake3Generator import IOQuake3Generator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controllers
    from configgen.Emulator import Emulator
    from configgen.types import Resolution


class TestIOQuake3Generator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[IOQuake3Generator]:
        return IOQuake3Generator

    @pytest.fixture
    def system_name(self) -> str:
        return 'quake3'

    @pytest.fixture
    def emulator(self) -> str:
        return 'ioquake3'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.create_file('/usr/bin/ioquake3/ioquake3', contents='ioquake3 bin')
        return fs

    @pytest.mark.parametrize(
        ('resolution', 'result'),
        [
            ({'width': 640, 'height': 480}, 4 / 3),
            ({'width': 1920, 'height': 1080}, 16 / 9),
            ({'width': 1920, 'height': 1144}, 16 / 9),
            ({'width': 1920, 'height': 1145}, 4 / 3),
        ],
        ids=str,
    )
    def test_get_in_game_ratio(  # pyright: ignore
        self, generator: IOQuake3Generator, resolution: Resolution, result: bool
    ) -> None:
        assert generator.getInGameRatio(SystemConfig({}), resolution, Path()) == result

    @pytest.mark.parametrize('core', ['ioquake3', 'vkquake3'])
    def test_generate(
        self,
        generator: IOQuake3Generator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(ROMS / 'quake3' / 'baseq3')
        fs.create_dir(ROMS / 'quake3' / 'missionpack')
        fs.create_file(ROMS / 'quake3' / 'Quake III Arena.quake3', contents='+set fs_game "baseq3"\nsome stuff\n')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'quake3' / 'Quake III Arena.quake3',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert filecmp.cmp('/usr/bin/ioquake3/ioquake3', ROMS / 'quake3' / 'ioquake3')
        assert (CONFIGS / 'ioquake3' / 'baseq3' / 'q3config.cfg').read_text() == snapshot(name='baseq3 config')
        assert (CONFIGS / 'ioquake3' / 'missionpack' / 'q3config.cfg').read_text() == snapshot(
            name='missionpack config'
        )

    @pytest.mark.parametrize(
        ('core', 'mock_system_config'),
        [
            ('ioquake3', {'ioquake3_mem': '64'}),
            ('vkquake3', {'ioquake3_mem': '64'}),
            ('vkquake3', {'vkquake3_api': 'opengl2'}),
            ('vkquake3', {'vkquake3_api': 'vulkan'}),
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: IOQuake3Generator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(ROMS / 'quake3' / 'baseq3')
        fs.create_file(ROMS / 'quake3' / 'Quake III Arena.quake3')

        generator.generate(
            mock_system,
            ROMS / 'quake3' / 'Quake III Arena.quake3',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'ioquake3' / 'baseq3' / 'q3config.cfg').read_text() == snapshot

    @pytest.mark.parametrize(
        ('core', 'mock_system_config'),
        [
            ('ioquake3', {}),
            ('ioquake3', {'ioquake3_mem': '64'}),
            ('vkquake3', {}),
            ('vkquake3', {'ioquake3_mem': '64'}),
            ('vkquake3', {'vkquake3_api': 'opengl2'}),
            ('vkquake3', {'vkquake3_api': 'vulkan'}),
        ],
        ids=str,
    )
    def test_generate_existing(
        self,
        generator: IOQuake3Generator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(ROMS / 'quake3' / 'ioquake3', contents='new ioquake3 bin')
        fs.create_dir(ROMS / 'quake3' / 'baseq3')
        fs.create_file(ROMS / 'quake3' / 'Quake III Arena.quake3')
        fs.create_file(
            CONFIGS / 'ioquake3' / 'baseq3' / 'q3config.cfg',
            contents="""seta r_mode "-1"
seta r_customwidth "640"
seta r_customheight "480"
seta in_joystickUseAnalog "0"
seta in_joystick "0"
seta cl_allowDownload "0"
seta com_hunkMegs "64"
seta cl_renderer "foo"
bind PAD0_LEFTSHOULDER "weapnext"
""",
        )

        generator.generate(
            mock_system,
            ROMS / 'quake3' / 'Quake III Arena.quake3',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (ROMS / 'quake3' / 'ioquake3').read_text() == 'new ioquake3 bin'
        assert (CONFIGS / 'ioquake3' / 'baseq3' / 'q3config.cfg').read_text() == snapshot(name='baseq3 config')
