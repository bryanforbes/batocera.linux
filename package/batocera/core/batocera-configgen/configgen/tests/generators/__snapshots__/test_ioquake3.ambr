# serializer version: 1
# name: TestIOQuake3Generator.test_generate[ioquake3]
  tuple(
    list([
      '/userdata/roms/quake3/ioquake3',
      '+set',
      'fs_game',
      '"baseq3"',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
    }),
  )
# ---
# name: TestIOQuake3Generator.test_generate[ioquake3][baseq3 config]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate[ioquake3][missionpack config]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate[vkquake3]
  tuple(
    list([
      '/userdata/roms/quake3/ioquake3',
      '+set',
      'fs_game',
      '"baseq3"',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
    }),
  )
# ---
# name: TestIOQuake3Generator.test_generate[vkquake3][baseq3 config]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "opengl2"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate[vkquake3][missionpack config]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "opengl2"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_config[ioquake3-{'ioquake3_mem': '64'}]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "64"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_config[vkquake3-{'ioquake3_mem': '64'}]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "64"
  seta cl_renderer "opengl2"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_config[vkquake3-{'vkquake3_api': 'opengl2'}]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "opengl2"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_config[vkquake3-{'vkquake3_api': 'vulkan'}]
  '''
  // generated by quake, do not modify
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "vulkan"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[ioquake3-{'ioquake3_mem': '64'}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "64"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[ioquake3-{}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[vkquake3-{'ioquake3_mem': '64'}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "64"
  seta cl_renderer "opengl2"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[vkquake3-{'vkquake3_api': 'opengl2'}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "opengl2"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[vkquake3-{'vkquake3_api': 'vulkan'}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "vulkan"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_generate_existing[vkquake3-{}][baseq3 config]
  '''
  seta r_mode "-1"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  seta com_hunkMegs "256"
  seta cl_renderer "opengl2"
  bind PAD0_LEFTSHOULDER "weapnext"
  seta r_customwidth "1920"
  seta r_customheight "1080"
  seta in_joystickUseAnalog "1"
  seta in_joystick "1"
  seta cl_allowDownload "1"
  bind PAD0_A "+moveup"
  bind PAD0_X "+movedown"
  bind PAD0_Y "+button2"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  
  '''
# ---
# name: TestIOQuake3Generator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'menu': list([
        'KEY_LEFTSHIFT',
        'KEY_ESC',
      ]),
    }),
    'name': 'ioquake3',
  })
# ---
